substrate-node/pallets/pallet-smart-contract/src/cost.rs

```diff
@@ -0,0 +1,265 @@
+    }
+
+    return total_cost.ceil().to_num::<u64>();
+}
+// cu1 = MAX(cru/2, mru/4)
+// cu2 = MAX(cru, mru/8)
+// cu3 = MAX(cru/4, mru/2)
+
+// CU = MIN(cu1, cu2, cu3)
+pub(crate) fn calculate_cu(cru: U64F64, mru: U64F64) -> U64F64 {
+    let mru_used_1 = mru / 4;
+    let cru_used_1 = cru / 2;
+    let cu1 = if mru_used_1 > cru_used_1 {
+        mru_used_1
+    } else {
+        cru_used_1
+    };
+
+    let mru_used_2 = mru / 8;
+    let cru_used_2 = cru;
+    let cu2 = if mru_used_2 > cru_used_2 {
+        mru_used_2
+    } else {
+        cru_used_2
+    };
+
+    let mru_used_3 = mru / 2;
+    let cru_used_3 = cru / 4;
+    let cu3 = if mru_used_3 > cru_used_3 {
+        mru_used_3
+    } else {
+        cru_used_3
+    };
+
+    let mut cu = if cu1 > cu2 { cu2 } else { cu1 };
+
+    cu = if cu > cu3 { cu3 } else { cu };
+
+    cu
+}
+
+// Calculates the discount that will be applied to the billing of the contract
+// Returns an amount due as balance object and a static string indicating which kind of discount it received
+// (default, bronze, silver, gold or none)
+pub fn calculate_discount<T: Config>(
+    amount_due: u64,
+    balance: BalanceOf<T>,
+    certification_type: NodeCertification,
+) -> (BalanceOf<T>, types::DiscountLevel) {
+    if amount_due == 0 {
+        return (
+            BalanceOf::<T>::saturated_from(0 as u128),
+            types::DiscountLevel::None,
+        );
+    }
+
+    let balance_as_u128: u128 = balance.saturated_into::<u128>();
+
+    // calculate amount due on a monthly basis
+    // we bill every one hour so we can infer the amount due monthly (30 days ish)
+    let amount_due_monthly = amount_due * 24 * 30;
+
+    // see how many months a user can pay for this deployment given his balance
+    let discount_level =
+        U64F64::from_num(balance_as_u128) / U64F64::from_num(amount_due_monthly);
+
+    // predefined discount levels
+    // https://wiki.threefold.io/#/threefold__grid_pricing
+    let discount_received = match discount_level.floor().to_num::<u64>() {
+        d if d >= 3 && d < 6 => types::DiscountLevel::Default,
+        d if d >= 6 && d < 12 => types::DiscountLevel::Bronze,
+        d if d >= 12 && d < 36 => types::DiscountLevel::Silver,
+        d if d >= 36 => types::DiscountLevel::Gold,
+        _ => types::DiscountLevel::None,
+    };
+
+    // calculate the new amount due given the discount
+    let mut amount_due = U64F64::from_num(amount_due) * discount_received.price_multiplier();
+
+    // Certified capacity costs 25% more
+    if certification_type == NodeCertification::Certified {
+        amount_due = amount_due * U64F64::from_num(1.25);
+    }
+
+    // convert to balance object
+    let amount_due: BalanceOf<T> =
+        BalanceOf::<T>::saturated_from(amount_due.ceil().to_num::<u64>());
+
+    (amount_due, discount_received)
+}
+
+pub fn calculate_cost_in_tft_from_musd<T: Config>(
+    total_cost: u64,
+) -> Result<u64, DispatchErrorWithPostInfo> {
+    // Todo fix me
```

Comment:
What is the fix me here? Maybe we can create a task for it?

Reply:
Fixed by erwan in #391